Sources/MapboxNavigation/WayNameLabel.swift

```diff
@@ -0,0 +1,154 @@
+           let legacyRoadShieldImage = legacyRoadShieldImage {
+            setAttributedText(roadName: roadName,
+                              roadShieldImage: legacyRoadShieldImage)
+            
+            return
+        }
+        
+        // For non US state road, use the generic shield icon first, then fall back to use the legacy shield.
+        if let roadShieldImage = spriteRepository.roadShieldImage(from: shieldRepresentation) {
+            setAttributedText(roadName: roadName,
+                              roadShieldImage: roadShieldImage,
+                              roadShieldText: shieldRepresentation?.text,
+                              roadShieldTextColor: shieldRepresentation?.textColor)
+            
+            return
+        }
+        
+        // In case if legacy shield icon is available - use it.
+        if let legacyRoadShieldImage = legacyRoadShieldImage {
+            setAttributedText(roadName: roadName,
+                              roadShieldImage: legacyRoadShieldImage)
+            
+            return
+        }
+        
+        // In case if neither generic nor legacy shield images are available - show only road name.
+        text = roadName
+    }
+    
+    private func setAttributedText(roadName: String,
+                                   roadShieldImage: UIImage,
+                                   roadShieldText: String? = nil,
+                                   roadShieldTextColor: String? = nil) {
+        let roadShieldAttributedString = roadShield(from: roadShieldImage,
+                                                    text: roadShieldText,
+                                                    textColor: roadShieldTextColor)
+        
+        let roadShieldAndNameAttributedString = NSMutableAttributedString()
+        // Road shield image is always drawn in front of the road name.
+        roadShieldAndNameAttributedString.append(roadShieldAttributedString)
+        roadShieldAndNameAttributedString.append(NSAttributedString(string: " \(roadName)"))
+        attributedText = roadShieldAndNameAttributedString
+    }
+    
+    private func roadShield(from image: UIImage,
+                            text: String? = nil,
+                            textColor: String? = nil) -> NSAttributedString {
+        let attachment = ShieldAttachment()
+        // Shield attachment should use similar font that is currently used in `WayNameLabel`.
+        attachment.font = font
+        // To correctly scale size of the font its height is based on the label where it is shown.
+        let fontSize = frame.size.height / 2.5
+        
+        // In case if `text` and `textColor` are valid - generate default road shield image, if not -
+        // generate legacy road shield image.
+        let roadShieldImage: UIImage
+        if let text = text,
+           let textColor = textColor {
+            let shieldColor = shieldColor(from: textColor)
```

Comment:
It's right. Because the `text` and `textColor` won't be `nil` if we could have the generic shield image. If they are invalid, that would go to the legacy one.